app/controllers/jobseekers/job_applications/build_controller.rb

```diff
@@ -4,59 +4,66 @@ class Jobseekers::JobApplications::BuildController < Jobseekers::ApplicationCont
 
   steps :personal_details, :professional_status, :personal_statement, :ask_for_support, :declarations
 
-  before_action :set_up_job_application, :set_up_current_step_number
-  before_action :set_up_show_form, only: %i[show]
-  before_action :set_up_update_form, only: %i[update]
+  helper_method :back_link_path, :current_step_number, :form, :job_application
 
   def show
     render_wizard
   end
 
   def update
-    application_data = @job_application.application_data.presence || {}
-
-    if @form.valid?
-      @job_application.assign_attributes(application_data: application_data.merge(form_params))
-      if params[:commit] == t("buttons.save_as_draft")
-        save_job_application_as_draft
-      else
-        render_wizard @job_application
-      end
+    if params[:commit] == t("buttons.save_as_draft")
+      save_job_application_as_draft
+    elsif form.valid?
+      job_application.assign_attributes(application_data: application_data.merge(form_params))
+      render_wizard job_application
     else
       render_wizard
     end
   end
 
   private
 
-  def form_params
-    send(FORM_PARAMS[step], params)
+  def application_data
+    @application_data ||= job_application.application_data.presence || {}
   end
 
-  def finish_wizard_path
-    jobseekers_job_application_review_path(@job_application)
+  def back_link_path
+    @back_link_path ||= case step
+                        when :personal_details
+                          new_jobseekers_job_application_path(job_application.vacancy.id)
+                        else
+                          previous_wizard_path
+                        end
   end
 
-  def save_job_application_as_draft
-    @job_application.save
-    redirect_to jobseekers_saved_jobs_path, success: t(".saved_job_application")
+  def current_step_number
+    @current_step_number ||= STEPS[step]
   end
 
-  def set_up_current_step_number
-    @current_step_number = STEPS[step]
+  def form
+    case action_name
```

Comment:
I reckon these might be better off staying as instance variables, but set in the individual actions – so it's obvious what they relate to.

i.e.

```ruby
def show
  @form = FORMS[step].new unless step == "wicked_finish"
  # ...
end
```